### PR TITLE
Change `asyncPropagating` default to true

### PR DESCRIPTION
**Pull request by tylerbenson**
Historically, we required setting a flag on the scope in order for async propagation to continue working.  This is a bad experience and a source of headaches.  With other changes in place, we are no longer as sensitive to over aggressive propagation, so we can change the default.  In the future we would like to remove this setting altogether.

### CODE REVIEW

File: dd-java-agent/instrumentation/elasticsearch/transport-2/src/latestDepTest/groovy/ElasticsearchBreakTraceInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.google.auto.service.AutoService;
+import datadog.trace.agent.tooling.Instrumenter;
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+
+@AutoService(Instrumenter.class)
+public class ElasticsearchBreakTraceInstrumentation implements Instrumenter {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/elasticsearch/transport-2/src/latestDepTest/groovy/ElasticsearchBreakTraceInstrumentation.java:12]: Could you add a comment to explain why this is necessary?

File: dd-java-agent/instrumentation/elasticsearch/transport-2/src/test/groovy/ElasticsearchBreakTraceInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.google.auto.service.AutoService;
+import datadog.trace.agent.tooling.Instrumenter;
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+
+@AutoService(Instrumenter.class)
+public class ElasticsearchBreakTraceInstrumentation implements Instrumenter {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/elasticsearch/transport-2/src/test/groovy/ElasticsearchBreakTraceInstrumentation.java:12]: Could you add a comment to explain how this differs from the other `ElasticsearchBreakTraceInstrumentation`?

File: dd-java-agent/instrumentation/elasticsearch/transport-5.3/src/test/groovy/ElasticsearchBreakTraceInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.google.auto.service.AutoService;
+import datadog.trace.agent.tooling.Instrumenter;
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+
+@AutoService(Instrumenter.class)
+public class ElasticsearchBreakTraceInstrumentation implements Instrumenter {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/elasticsearch/transport-5.3/src/test/groovy/ElasticsearchBreakTraceInstrumentation.java:12]: Could you add a comment to explain how this differs from the other `ElasticsearchBreakTraceInstrumentation`?

File: dd-java-agent/instrumentation/elasticsearch/transport-5/src/test/groovy/ElasticsearchBreakTraceInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.google.auto.service.AutoService;
+import datadog.trace.agent.tooling.Instrumenter;
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+
+@AutoService(Instrumenter.class)
+public class ElasticsearchBreakTraceInstrumentation implements Instrumenter {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/elasticsearch/transport-5/src/test/groovy/ElasticsearchBreakTraceInstrumentation.java:12]: Could you add a comment to explain how this differs from the other `ElasticsearchBreakTraceInstrumentation`?

File: dd-java-agent/instrumentation/elasticsearch/transport-6/src/test/groovy/ElasticsearchBreakTraceInstrumentation.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.google.auto.service.AutoService;
+import datadog.trace.agent.tooling.Instrumenter;
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+
+@AutoService(Instrumenter.class)
+public class ElasticsearchBreakTraceInstrumentation implements Instrumenter {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/elasticsearch/transport-6/src/test/groovy/ElasticsearchBreakTraceInstrumentation.java:12]: Could you add a comment to explain how this differs from the other `ElasticsearchBreakTraceInstrumentation`?

File: dd-java-agent/instrumentation/java-concurrent/src/test/groovy/NettyExecutorInstrumentationTest.groovy [hunk 6]
```diff
@@ -14,7 +14,6 @@ import java.util.concurrent.Future
 import java.util.concurrent.RejectedExecutionException
 import java.util.concurrent.TimeUnit
 
-import static datadog.trace.bootstrap.instrumentation.api.AgentTracer.activeScope
 import static org.junit.Assume.assumeTrue
 
 class NettyExecutorInstrumentationTest extends AgentTestRunner {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/java-concurrent/src/test/groovy/NettyExecutorInstrumentationTest.groovy:19]: How do we avoid diffs like this? One option is for you to uninstall the plugin which edits code unrelated to the change you've made when you save files as I believe you are now in a minority in using it. If we want consistent formatting of test files, note that I run `./gradlew spotlessApply` before I commit every time I commit, but reformatting of test files is disabled; I think it should have been enabled. The status quo is unsustainable.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/AdviceUtils.java [hunk 7]
```diff
@@ -22,9 +22,7 @@
     if (state != null) {
       final TraceScope.Continuation continuation = state.getAndResetContinuation();
       if (continuation != null) {
-        final TraceScope scope = continuation.activate();
```
Comment (dougqh) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/AdviceUtils.java:25]: I wasn't expected the set to true blocks to be removed in this first step.
I was assuming we just set async propagation will to true and that's it.

**Review by dougqh — CHANGES_REQUESTED** [review]
As first step, I would prefer we set async propagation to true, but not remove the places in the code where we explicitly set to true.

I would assume that the default behavior would then be semantically equivalent to this.
However, not removing the set(true) would give us to override the default and return to the prior behavior.

File: dd-java-agent/instrumentation/opentracing/api-0.31/src/test/groovy/OpenTracing31Test.groovy [hunk 8]
```diff
@@ -131,6 +131,7 @@ class OpenTracing31Test extends AgentTestRunner {
     setup:
     def span = tracer.buildSpan("some name").start()
     def scope = tracer.scopeManager().activate(span, finishSpan)
+    (scope as TraceScope).setAsyncPropagation(false)
```
[the remark this answers is not in the document]
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentracing/api-0.31/src/test/groovy/OpenTracing31Test.groovy:134]: This was not a new gotcha... it was returning the state to false so the rest of the test could remain the same.  I could just as easily invert the test.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/AdviceUtils.java (hunk 7 above)
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/AdviceUtils.java]: I've reverted those changes and added a flag we can flip to change the default again.

File: dd-java-agent/instrumentation/opentracing/api-0.31/src/test/groovy/OpenTracing31Test.groovy (hunk 8 above)
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/instrumentation/opentracing/api-0.31/src/test/groovy/OpenTracing31Test.groovy:134]: I'm picking up on why it's necessary to make this change at all - won't this break a lot of manually traced applications if it's strictly necessary to invert the flag?

[the remark this answers is not in the document]
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentracing/api-0.31/src/test/groovy/OpenTracing31Test.groovy:134]: Previously we were relying on the implicit default state. In the `expect` block below, we are explicitly asserting:
```
    !(scope as TraceScope).isAsyncPropagating()
    (scope as TraceScope).capture() == null
```

This is not something a customer would likely be doing.

Changing the default requires this test to change.  Would you prefer continuing to rely on the default and change the asserts or make the expected state implicit up front?

**Review by richardstartin — APPROVED** [review]

**Review by dougqh — APPROVED** [review]

**Pull request merged**